Libraries/LibJS/Runtime/Iterator.h

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2020, The SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Function.h>
+#include <LibJS/Runtime/Object.h>
+
+namespace JS {
+
+class Iterator final : public Object {
+public:
+    struct IteratorResult {
+        bool finished { false };
+        Value value;
+    };
+
+    Iterator(Object* iterable, AK::Function<IteratorResult(Object*, const Vector<Value>&)>&& next_function);
```

Comment:
Wait, is `iterable` ever `nullptr`? If not, let's use `Object&`.

Reply:
> Okay I think you'll have to give a high-level overview of what's going on here.

Here goes.
The function is parsed into a GenratorFunction, which holds the original AST nodes.
At each call to the function, we create a new `GeneratorFunction` (this is due to GeneratorFunction being both an Iterable and an Iterator, I blame js), and then destructively go through its statements.
Once we hit a return, or run out of nodes, the generator is considered to have finished execution.

> Why do we need to duplicate the AST children into a temporary block statement and remove them as we go?

We would otherwise need to keep some sort of instruction pointer, which is something that our Interpreter does not have the facilities to handle. 

> What is the lexical environment allocation strategy?
Allocated once per call to a generator function, if needed.
reused for the entire lifetime of the generator instance

Reply:
...wrong place